### ISSUE TITLE
Add unit tests

### ISSUE THREAD

**Issue by thompsy** [Ewan10/Task-Manager#5]
Currently there are no unit tests. We should definitely be doing some testing.

You may need to think about refactoring some of your methods to make them easier to test but we can talk about that in person. In the mean-time here's some reasonable discussion about some high-level design principles that can make code easier to test: https://softwareengineering.stackexchange.com/questions/153410/what-are-the-design-principles-that-promote-testable-code-designing-testable-c

**Comment by Ewan10** [comment]
Yes those principles SOLID are presented in Urma's, Warburton's book. I printed out some other notes you sent some months ago about testing from Baeldung site or so and I read them. I let you arrange a meetup to discuss how to write tests for the public API as I can see multiple ways and some may not conform to the formal testing standards. I thought about refactoring the original methods but I can think of other ways which might be inappropriate.

**Comment by thompsy** [comment]
Once you've finished that current PR you could start on this. The `TaskManager` will be factored in a way that makes it easier to test. There are a number of things we could test:

- the standard happy flow of loading a file and saving it and comparing that the two files are identical
- the happy case where we update the status of a task and check we save it correctly
- the unhappy case where the input file that we're loading is badly formatted or has duplicate ids.
- the unhappy case where the file doesn't exist

In each case a sensible approach would be to have a separate test for each of these cases. Each test case should probably have an input file and an output file that live the `src/test/resources/` folder. Each test case can then create a `TaskManager` class, load the input file, take some action e.g. update the status of a task, save the output file to a temporary location and compare it with the expected output file.

**Comment by Ewan10** [comment]
Andrew after I merged the PR the additions of the test code are available in the main branch with the test folder and you can open the TaskManagerTest.java and see where I stuck. I don't know how to translate some tests to code. If you could have a look?

**Comment by thompsy** [comment]
Ok, so two quick things here. 

Firstly, you should never merge code straight into the `main` branch. All code that goes into the `main` branch should start in it's own separate branch and then be reviewed in a PR before being merged. A PR is the place where we should be discussing these questions rather than here as it's much harder to comment on specific lines of code here.

Secondly, can you outline here precisely what tests you're having trouble with. What is the test name? What are you trying to test? What will the inputs be? What will the expected outputs be?

**Comment by Ewan10** [comment]
I 'll have to use some particular dependency and introduce it through Maven for sqlite. Searching in the Maven repository I found that link: [ https://mvnrepository.com/artifact/org.xerial/sqlite-jdbc](url) You didn't specify any repository I don't know which you want me to employ. The video tutorials I watched on youtube show sqlite usage but for environments like eclipse and I am not sure if the files for sqlite are identical with this link. I just want to search for the documentation to read and then ask you for guidance. 

One question I have is I 'll have to update the save and load methods for operations in the database so by creating a new local branch sqlite-as-backend will I overwrite the code of save and load methods? If Idelete the refactor-user-input branch as I merged the changes the older version of the code where I load and save in a file how will be accessible/available?

**Comment by Ewan10** [comment]
Just above you specified the tests for me to implement: 

1.     the standard happy flow of loading a file and saving it and comparing that the two files are identical
2.     the happy case where we update the status of a task and check we save it correctly
3.     the unhappy case where the input file that we're loading is badly formatted or has duplicate ids.
4.     the unhappy case where the file doesn't exist

So inside main you will find a test folder which I uploaded for the first time so it is not for this time hard to track the new code. In test/java it is the only file TaskManager.java There is where I implemented the tests by test methods. It is one class the TaskManagerTest.java and it is implemented one method per test inside there.  Inside the method body where I didn't implement the code I wrote comments of what should be done but I find it uncertain how to code it. This is for test 3 and test 1. 

How to check if the file is badly formatted? If it has duplicate ids I need to write a method that accesses and compares the id's of the tasks in the map. Should I add this method in the TaskManager.java class? As I need to access fields of tasks. I see that I need to create methods and extend the interface for accessing fields to perform the tests and Idon't know where to define them.

Regarding test 1 you can see the comments inside the body of the method in TaskManagerTest.java
 `What are the exact objects in assertEquals() to be compared?
         * Assertions.assertEquals(expected, actual);` 
      
   Should I compare tasks using a loop? Or should I compare the maps?
Also I imported the library junit.jupiter. Is this the one you intended to use?

**Comment by Ewan10** [comment]
Just above you specified the tests for me to implement: 

1.     the standard happy flow of loading a file and saving it and comparing that the two files are identical
2.     the happy case where we update the status of a task and check we save it correctly
3.     the unhappy case where the input file that we're loading is badly formatted or has duplicate ids.
4.     the unhappy case where the file doesn't exist

So inside main you will find a test folder which I uploaded for the first time so it is not for this time hard to track the new code. In test/java it is the only file TaskManager.java There is where I implemented the tests by test methods. It is one class the TaskManagerTest.java and it is implemented one method per test inside there.  Inside the method body where I didn't implement the code I wrote comments of what should be done but I find it uncertain how to code it. This is for test 3 and test 1. 

How to check if the file is badly formatted? If it has duplicate ids I need to write a method that accesses and compares the id's of the tasks in the map. Should I add this method in the TaskManager.java class? As I need to access fields of tasks. I see that I need to create methods and extend the interface for accessing fields to perform the tests and Idon't know where to define them.

Regarding test 1 you can see the comments inside the body of the method in TaskManagerTest.java
 `What are the exact objects in assertEquals() to be compared?
         * Assertions.assertEquals(expected, actual);` 
      
   Should I compare tasks using a loop? Or should I compare the maps?
Also I imported the library junit.jupiter. Is this the one you intended to use?

**Comment by thompsy** [comment]
So, writing these tests should serve as a prompt for you to think further about the functionality of the program that you're writing and how it should behave under certain circumstances.

So, for example, how does it handle a badly formatted file? There are a number of ways a file could be badly formatted. It could have duplicate ids, it could be invalid json, it could be empty, it could be full of garbage data. How should your program handle each of these cases? How would it know? Would it just throw an exception, or would it read in the data and try and use it?

In terms of the first test and what you should be comparing, perhaps the best way to think about this is to think about what state the `TaskManager` is holding. This is basically just the `tasks` map and the `id`. You should probably compare both of these to expected values to make sure the state is correct. Whether you do that in a loop or something else isn't hugely important at this stage. The key thing is that your code is correct and easy to understand.

**Comment by Ewan10** [comment]
Andrew don't you have a look to the tests in add-unit-tests branch and make some comments?
